from functools import wraps
from database import Session
from sqlalchemy.exc import DataError, OperationalError, ProgrammingError, DatabaseError, SQLAlchemyError, IntegrityError

import logging
from utils.logging_config import setup_logger

# Configura o logger
setup_logger()

# Decorador para gerenciar sess√µes e tratar exce√ß√µes
def session_manager(commit=True):  # Par√¢metro para definir se deve dar commit
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            with Session() as session:
                try:
                    result = func(*args, session=session, **kwargs)  # Executa a fun√ß√£o

                    if commit:
                        session.commit()
                        logging.info(f"Commit realizado com sucesso na fun√ß√£o '{func.__name__}'.")

                    # üîπ Evita aninhamento extra: se j√° for um dicion√°rio com "success", retorna diretamente
                    if isinstance(result, dict) and "success" in result:
                        return result

                    return {"success": True, "data": result}  # Apenas encapsula se necess√°rio

                except DataError as e:
                    session.rollback()
                    logging.error(f"Erro de formato de dado inv√°lido na fun√ß√£o '{func.__name__}': {e}", exc_info=True)
                    return {"success": False, "error": "Erro: Formato de dado inv√°lido ou valor fora do intervalo permitido."}

                except IntegrityError as e:
                    session.rollback()
                    logging.error(f"Erro de integridade: {e}", exc_info=True)
                    return {"success": False, "error": "Erro: Viola√ß√£o de integridade. Registro duplicado ou dados inv√°lidos."}

                except OperationalError as e:
                    session.rollback()
                    logging.error(f"Erro operacional: {e}", exc_info=True)
                    return {"success": False, "error": "Erro operacional: Falha na conex√£o com o banco de dados."}

                except Exception as e:
                    session.rollback()
                    logging.error(f"Erro inesperado '{func.__name__}': {e}", exc_info=True)
                    return {"success": False, "error": f"Erro inesperado: {e}"}

        return wrapper
    return decorator

